contracts/BSCValidatorSet.sol

```diff
@@ -468,17 +556,175 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
       }
     }
     if (m>n) {
+      ValidatorExtra memory validatorExtra;
       for (uint i = n;i<m;i++) {
         currentValidatorSet.push(validatorSet[i]);
+        validatorExtraSet.push(validatorExtra);
         currentValidatorSetMap[validatorSet[i].consensusAddress] = i+1;
       }
     }
+
+    // make sure all new validators are cleared maintainInfo
+    // should not happen, still protect
+    numOfMaintaining = 0;
+    for (uint i = 0; i < currentValidatorSet.length; i++) {
+      validatorExtraSet[i].isMaintaining = false;
+      validatorExtraSet[i].enterMaintenanceHeight = 0;
+    }
   }
 
   function isSameValidator(Validator memory v1, Validator memory v2) private pure returns(bool) {
     return v1.consensusAddress == v2.consensusAddress && v1.feeAddress == v2.feeAddress && v1.BBCFeeAddress == v2.BBCFeeAddress && v1.votingPower == v2.votingPower;
   }
 
+  function _misdemeanor(address validator) private returns (uint256) {
+    uint256 index = currentValidatorSetMap[validator];
+    if (index <= 0) {
+      return ~uint256(0);
+    }
+    // the actually index
+    index = index - 1;
+
+    uint256 income = currentValidatorSet[index].incoming;
+    currentValidatorSet[index].incoming = 0;
+    uint256 rest = currentValidatorSet.length - 1;
+    emit validatorMisdemeanor(validator, income);
+    if (rest == 0) {
+      // should not happen, but still protect
+      return index;
+    }
+    uint256 averageDistribute = income / rest;
+    if (averageDistribute != 0) {
+      for (uint i = 0; i < index; i++) {
+        currentValidatorSet[i].incoming = currentValidatorSet[i].incoming + averageDistribute;
+      }
+      uint n = currentValidatorSet.length;
+      for (uint i = index + 1; i < n; i++) {
+        currentValidatorSet[i].incoming = currentValidatorSet[i].incoming + averageDistribute;
+      }
```

Comment:
Suggest this format:
for (uint i = 0; i < currentValidatorSet.length; i++) {
    if (i == index)
        continue;

    currentValidatorSet[i].incoming = currentValidatorSet[i].incoming + averageDistribute;
}

Reply:
This is old code, and I try not to change the old code if it is not necessary